sphinx/ext/autodoc/__init__.py

```diff
@@ -1188,22 +1196,72 @@ def format_args(self, **kwargs: Any) -> str:
         if self.env.config.autodoc_typehints in ('none', 'description'):
             kwargs.setdefault('show_annotation', False)
 
-        # for classes, the relevant signature is the __init__ method's
-        initmeth = self.get_attr(self.object, '__init__', None)
-        # classes without __init__ method, default __init__ or
-        # __init__ written in C?
-        if initmeth is None or \
-                inspect.is_builtin_class_method(self.object, '__init__') or \
-                not(inspect.ismethod(initmeth) or inspect.isfunction(initmeth)):
-            return None
-        try:
-            self.env.app.emit('autodoc-before-process-signature', initmeth, True)
-            sig = inspect.signature(initmeth, bound_method=True)
+        def get_user_defined_function_or_method(obj: Any, attr: str) -> Any:
+            """ Get the `attr` function or method from `obj`, if it is user-defined. """
+            if inspect.is_builtin_class_method(obj, attr):
+                return None
+            attr = self.get_attr(obj, attr, None)
+            if not (inspect.ismethod(attr) or inspect.isfunction(attr)):
+                return None
+            return attr
+
+        sig = None
+
+        # this sequence is copied from inspect._signature_from_callable
+
+        # First, let's see if it has an overloaded __call__ defined
+        # in its metaclass
+        if sig is None:
+            call = get_user_defined_function_or_method(type(self.object), '__call__')
+
+            if call is not None:
+                if "{0.__module__}.{0.__qualname__}".format(call) in _METACLASS_CALL_BLACKLIST:
+                    call = None
+
+            if call is not None:
+                self.env.app.emit('autodoc-before-process-signature', call, True)
+                try:
+                    sig = inspect.signature(call, bound_method=True)
+                except ValueError:
```

Comment:
I think we need to handle both `ValueError` and `TypeError`.

Reply:
Can you come up with an example that raises `TypeError`?

Reply:
I got an error with this class.
```
class Foo:
    __signature__ = 1
```
```
WARNING: error while formatting arguments for example.Foo:
Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 395, in format_signature
    args = self.format_args(**kwargs)
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 1255, in format_args
    sig = inspect.signature(self.object, bound_method=False)
  File "/Users/tkomiya/work/sphinx/sphinx/util/inspect.py", line 406, in signature
    signature = inspect.signature(subject)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 3093, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2842, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2252, in _signature_from_callable
    raise TypeError(
TypeError: unexpected object 1 in __signature__ attribute

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 398, in format_signature
    args = self.format_args()
  File "/Users/tkomiya/work/sphinx/sphinx/ext/autodoc/__init__.py", line 1255, in format_args
    sig = inspect.signature(self.object, bound_method=False)
  File "/Users/tkomiya/work/sphinx/sphinx/util/inspect.py", line 406, in signature
    signature = inspect.signature(subject)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 3093, in signature
    return Signature.from_callable(obj, follow_wrapped=follow_wrapped)
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2842, in from_callable
    return _signature_from_callable(obj, sigcls=cls,
  File "/Users/tkomiya/.pyenv/versions/3.8.2/lib/python3.8/inspect.py", line 2252, in _signature_from_callable
    raise TypeError(
TypeError: unexpected object 1 in __signature__ attribute
```

I know this is crazy example. But I received such report sometimes. So it would be better to handle `ValueErorr` also.

Reply:
I found related problem in autofunction: https://github.com/sphinx-doc/sphinx/issues/7629

Reply:
I found a real bug report: https://github.com/sphinx-doc/sphinx/issues/7138

Reply:
That real bug report is `ValueError`, which is already handled.

I think I should try to make that traceback half as long though, pushing an update now.

Reply:
Oh, sorry. The last one is my wrong.

Reply:
>>Please catch a TypeError in format_args(). Then I'll merge this soon.
>
>Would you mind adding a comment on the relevant lines? I've moved things around a bit now.

Sorry. As commented in this thread, `inspect.signature()` raises `TypeError`. And I think `format_args()` should not pass through it. So please catch it like `FunctionDocumenter.format_args()`.

Reply:
Thanks for the reminder, updated.

Reply:
Thank you for update. I'll review this tomorrow (or in this weekend) with cleared eyes.